hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java

```diff
@@ -373,6 +373,7 @@ private OzoneConsts() {
 
   // For Multipart upload
   public static final int OM_MULTIPART_MIN_SIZE = 5 * 1024 * 1024;
+  public static final int OM_MULTIPART_MAX_PART_NUMBER = 10000;
```

Comment:
@xichen01 , If there are only `10,000` parts, the memory optimization may not be significant.  What is the memory gain for this change?

Reply:
## Test Result
**The memory usage after 10000 "part key" upload.**

```shell
[root@centos /root/]% jmap -histo:live `jps | grep OzoneManagerStarter | awk '{ print $1 }'` | head -n 20
 num     #instances         #bytes  class name
----------------------------------------------
   1:      46497817     1859912680  java.util.TreeMap$Entry
   1:      13112075      419586400  java.util.HashMap$Node
   2:        558238       78747928  [B
   3:        444650       34326104  [C
   4:        268746       14390832  [Ljava.lang.Object;
   5:        260101       12484848  org.apache.hadoop.hdds.protocol.proto.HddsProtos$Port
   6:        486330       11671920  com.google.protobuf.LiteralByteString
   7:        443954       10654896  java.lang.String
   8:        330385       10572320  org.apache.hadoop.ozone.OzoneAcl
   9:        330434        7932848  [J
  10:        330388        7929312  java.util.BitSet
  11:        294524        7068576  java.util.ArrayList
  12:        104041        5826296  org.apache.hadoop.hdds.protocol.proto.HddsProtos$UUID
  13:         22023        5814072  org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos$OMResponse
  14:         72411        5639624  [Ljava.util.HashMap$Node;
  15:        220320        5287680  org.apache.hadoop.hdds.protocol.DatanodeDetails$Port
  16:        101153        4855344  java.util.HashMap
  17:        148143        4740576  java.util.UUID
```
**After the upload of 10000 parts, 2G of `ParOldGen` memory will be generated**

```shell
[root@centos /root/]%   jcmd `jps | grep OzoneManagerStarter | awk '{ print $1 }'` GC.heap_info
33896:
PSYoungGen total 691712K, used 448768K [0x000000076ab00000, 0x000000079df80000, 0x00000007c0000000)
eden space 566784K, 78% used [0x000000076ab00000,0x0000000785faef78,0x000000078d480000)
from space 124928K, 1% used [0x0000000796580000,0x0000000796711278,0x000000079df80000)
to space 136704K, 0% used [0x000000078d480000,0x000000078d480000,0x0000000795a00000)
ParOldGen total 2796544K, used 2115856K [0x00000006c0000000, 0x000000076ab00000, 0x000000076ab00000)
object space 2796544K, 75% used [0x00000006c0000000,0x0000000741244250,0x000000076ab00000)
Metaspace used 57872K, capacity 60290K, committed 60800K, reserved 1103872K
class space used 6552K, capacity 6955K, committed 7040K, reserved 1048576K
```
---
## Reson
get OmMultipartKeyInfo copies out a new copy each time, so 10000 is only one memory consumption

get OmMultipartKeyInfo 
https://github.com/apache/ozone/blob/cffa3864b974499078fbdb37506ed992319b09c5/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/multipart/S3MultipartUploadCommitPartRequest.java#L149-L150

copies out a new copy
https://github.com/apache/ozone/blob/cffa3864b974499078fbdb37506ed992319b09c5/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java#L176

**So, If we have 10,000 parts when multiupload, there will create (1 + 10000) * 10000 / 2 = 50005000 `TreeMap.Entry`**

Reply:
@xichen01 , I see this is the same O(n^2) problem we found in HDDS-8238 .  We should fix it by changing the RocksDb table schema; see "3. Change the RocksDb table schema such as below:" in the first comment of HDDS-8238.

Since this fix require enforcing `OM_MULTIPART_MAX_PART_NUMBER`, it is an incompatible change.  Are you interested in working on HDDS-8238?

Reply:
> Since this fix require enforcing `OM_MULTIPART_MAX_PART_NUMBER`, it is an incompatible change.

I'm not sure about that.  We already have this limitation in `createMultipartKey` on `master`:

https://github.com/apache/ozone/blob/cffa3864b974499078fbdb37506ed992319b09c5/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java#L1713-L1714

Reply:
`OM_MULTIPART_MAX_PART_NUMBER` has already been applied before, no more than 10000 parts are allowed to be uploaded.
enforcing OM_MULTIPART_MAX_PART_NUMBER is a redundancy constraint to prevent any unexpected situation that would cause `IntSparseMap` to insert too many elements and waste memory.
```java
    public Builder() {
      this.partKeyInfoList =
          new IntSparseMap<>(EMPTY_PART_KEY_INFO, OM_MULTIPART_MAX_PART_NUMBER);
    }
```

Reply:
@szetszwo If we change RocksDB, it may not be fully compatible with the previous version, e.g. upgrade and roll back to the old version (Correct me if I'm wrong)
This PR change is fully compatible with the old version, but does not fix the RocksDB issue, so there may be a trade-off here

Reply:
@adoroszlai , thanks for pointing out the limit in the RpcClient.  Not sure why we have it in the first place.  If we use Amazon client, then it should be able to exceed the 10k limit.

With this change, it will limit everywhere using `OmMultipartKeyInfo`.

Reply:
@xichen01 , we should make it compatible -- keep our OM code for the previous record and add new code for the change.  We may consider doing it in a different table.  That will fix source of the problem.

Reply:
@szetszwo 
I have a compatibility issue b don't know how to solve it
Suppose we have two versions VOLD and VNEW, VNEW will use a new RocksDB table TNEW for `multiupload`, VOLD uses TOLD

The user wants to upload 100 slices. Now the user uploads 50 in VNEW, then the cluster rolls back to VOLD, but VOLD does not have the logic to handle the `xxKeyInfo` in the new table VNEW, VOLD will not know that the user has finished uploading 50 slices, which may have compatibility issues.

To achieve compatibility in the above scenario, it may be necessary to write both TOLD and TNEW in the VNEW version, but the issue caused by TOLD is still not solved

Reply:
In fact even AWS has the same limit.

> Part numbers can be any number from 1 to 10,000, inclusive.

https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html

> Maximum number of parts per upload: 10,000

https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html

Reply:
I see.  Thanks for the info.  We probably should document it if we haven't.

Reply:
> @szetszwo I have a compatibility issue b don't know how to solve it Suppose we have two versions VOLD and VNEW, VNEW will use a new RocksDB table TNEW for `multiupload`, VOLD uses TOLD
> 
> The user wants to upload 100 slices. Now the user uploads 50 in VNEW, then the cluster rolls back to VOLD, but VOLD does not have the logic to handle the `xxKeyInfo` in the new table VNEW, VOLD will not know that the user has finished uploading 50 slices, which may have compatibility issues.
> 
> To achieve compatibility in the above scenario, it may be necessary to write both TOLD and TNEW in the VNEW version, but the issue caused by TOLD is still not solved

Is there a good solution to the RocksDB log issue, without breaking compatibility, or do we need to tolerate some minor incompatibilities